src/core/components/pin.js

```diff
@@ -0,0 +1,563 @@
+'use strict'
+
+const dagPB = require('ipld-dag-pb')
+const DAGNode = dagPB.DAGNode
+const DAGLink = dagPB.DAGLink
+const CID = require('cids')
+const pinSet = require('./pin-set')
+const promisify = require('promisify-es6')
+const multihashes = require('multihashes')
+const Key = require('interface-datastore').Key
+const _ = require('lodash')
+const each = require('async/each')
+const waterfall = require('async/waterfall')
+const until = require('async/until')
+
+const keyString = multihashes.toB58String
+function KeySet (keys) {
+  // Buffers with identical data are still different objects, so
+  // they need to be cast to strings to prevent duplicates in Sets
+  this.keys = {}
+  this.add = (key) => {
+    this.keys[keyString(key)] = key
+  }
+  this.delete = (key) => {
+    delete this.keys[keyString(key)]
+  }
+  this.clear = () => {
+    this.keys = {}
+  }
+  this.has = (key) => {
+    return (keyString(key) in this.keys)
+  }
+  this.toArray = () => {
+    return Object.keys(this.keys).map((hash) => {
+      return this.keys[hash]
+    })
+  }
+  this.toStringArray = () => {
+    return Object.keys(this.keys)
+  }
+  keys = keys || []
+  keys.forEach(this.add)
+}
+
+module.exports = function pin (self) {
+  let directPins = new KeySet()
+  let recursivePins = new KeySet()
+  let internalPins = new KeySet()
+
+  // temp - should use /local/pins to be consistent with go repo, but
+  // the pin set in datastore in test/go-ipfs-repo doesn't deserialize
+  const pinDataStoreKey = new Key('/local/pins/js')
+
+  const repo = self._repo
+  const dag = self.dag
+
+  function normalizeHashes (hashes, callback) {
+    // try to accept a variety of hash options including
+    // multihash Buffers, base58 strings, and ipfs path
+    // strings, either individually or as an array
+    if (Buffer.isBuffer(hashes) || typeof hashes.forEach !== 'function') {
+      hashes = [hashes]
+    }
+    const normalized = {
+      hashes: [],
+      update: (multihash, cb) => {
+        try {
+          multihashes.validate(multihash)
+        } catch (err) { return cb(err) }
+
+        normalized.hashes.push(multihash)
+        cb()
+      }
+    }
+    each(hashes, (hash, cb) => {
+      if (typeof hash === 'string') {
+        // example: '/ipfs/QmRootHash/links/by/name'
+        const matched = hash.match(/^(?:\/ipfs\/)?([^/]+(?:\/[^/]+)*)\/?$/)
+        if (!matched) {
+          return cb(new Error('invalid ipfs ref path'))
+        }
```

Comment:
You can use http://npmjs.org/is-ipfs for this one.

Reply:
I was also using the regex to capture and split the part of the path after "ipfs/" to get the individual links in the path, is there an import I can use for that?